geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/api/package.scala

```diff
@@ -325,4 +328,54 @@ package object api {
   case class LowerBoundedRange[T](lower: T) extends ScanRange[T]
   case class UpperBoundedRange[T](upper: T) extends ScanRange[T]
   case class UnboundedRange[T](empty: T) extends ScanRange[T]
+
+  /**
+    * Trait for simple serialization that can be defined using a Map of strings. Implementations must have
+    * a zero-arg constructor to allow for instantiation via reflection
+    */
+  trait SerializableState {
+
+    /**
+      * Initialize an instance after deserialization
+      *
+      * @param state state
+      */
+    def init(state: Map[String, String]): Unit
+
+    /**
+      * The state used to serialize this instance
+      *
+      * @return
+      */
+    def state: Map[String, String]
+  }
+
+  object SerializableState {
+
+    /**
+      * Serialize an object
+      *
+      * @param obj object to serialize
+      * @return
+      */
+    def serialize(obj: SerializableState): String =
+      s"${obj.getClass.getName},${StringSerialization.encodeMap(obj.state)}"
```

Comment:
Sanity check:  Does the map have more commas in it?

Reply:
It does, but we only split on the first comma to get the class name, then pass the rest back to `StringSerialization.decodeMap` which handles them appropriately.